api/v1/discovery_webhook.go

```diff
@@ -0,0 +1,148 @@
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"errors"
+	"fmt"
+
+	admissionregistration "k8s.io/api/admissionregistration/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	cl "sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// log is for logging in this package.
+var (
+	discoveredclusterLog = logf.Log.WithName("discoveredcluster-resource")
+	Client               cl.Client
+
+	ErrInvalidImportStrategy = errors.New("invalid import-strategy")
+)
+
+// ValidatingWebhook returns the ValidatingWebhookConfiguration used for the discoveredcluster
+// linked to a service in the provided namespace
+func ValidatingWebhook(namespace string) *admissionregistration.ValidatingWebhookConfiguration {
+	fail := admissionregistration.Fail
+	none := admissionregistration.SideEffectClassNone
+	path := "/validate-discovery-open-cluster-management-io-v1-discoveredcluster"
+	return &admissionregistration.ValidatingWebhookConfiguration{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "admissionregistration.k8s.io/v1",
+			Kind:       "ValidatingWebhookConfiguration",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "discovery.open-cluster-management.io",
+			Annotations: map[string]string{"service.beta.openshift.io/inject-cabundle": "true"},
+		},
+		Webhooks: []admissionregistration.ValidatingWebhook{
+			{
+				AdmissionReviewVersions: []string{
+					"v1",
+					"v1beta1",
+				},
+				Name: "discovery.open-cluster-management.io",
+				ClientConfig: admissionregistration.WebhookClientConfig{
+					Service: &admissionregistration.ServiceReference{
+						Name:      "discovery-operator-webhook-service",
+						Namespace: namespace,
+						Path:      &path,
+					},
+				},
+				FailurePolicy: &fail,
+				Rules: []admissionregistration.RuleWithOperations{
+					{
+						Rule: admissionregistration.Rule{
+							APIGroups:   []string{GroupVersion.Group},
+							APIVersions: []string{GroupVersion.Version},
+							Resources:   []string{"discoveredclusters"},
+						},
+						Operations: []admissionregistration.OperationType{
+							admissionregistration.Create,
+							admissionregistration.Update,
+							admissionregistration.Delete,
+						},
+					},
+				},
+				SideEffects: &none,
+			},
+		},
+	}
+}
+
+func (r *DiscoveredCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	Client = mgr.GetClient()
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+var _ webhook.Defaulter = &DiscoveredCluster{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *DiscoveredCluster) Default() {
+	discoveredclusterLog.Info("default", "Name", r.Name)
+}
+
+var _ webhook.Validator = &DiscoveredCluster{}
```

Comment:
Why is there a global variable that's not defined?

Reply:
Following the `kubebuilder` docs, they were adding that line when they were building the `webhook`: https://book.kubebuilder.io/multiversion-tutorial/webhooks.html#and-maingo. I checked out a few other repos that are configuring a webhook and they are using the same logic, so it might be best for us to remain consistent with the other projects/components.